File: src/main/java/ee/leola/kassa/models/Transaction.java
package ee.leola.kassa.models;

import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import java.math.BigDecimal;
import java.time.Instant;
import java.time.format.DateTimeFormatter;
import java.util.Set;
import lombok.Data;
import lombok.EqualsAndHashCode;

@Entity
@Data
@EqualsAndHashCode(callSuper = true)
public class Transaction extends Model {

  private boolean invalid;

  private Instant created = Instant.now();

  @NotNull @ManyToOne private User user;

  @OneToMany(cascade = CascadeType.ALL)
  @JoinTable(
      name = "transaction_transaction_product",
      joinColumns = @JoinColumn(name = "transaction_id", referencedColumnName = "id"),
      inverseJoinColumns =
          @JoinColumn(name = "transaction_product_id", referencedColumnName = "id"))
  private Set<TransactionProduct> products;

  @NotNull @ManyToOne private Paytype paytype;

  public String getFormattedTime() {
    return DateTimeFormatter.ISO_INSTANT.format(created);
  }

  public BigDecimal getSum() {
    BigDecimal sum = BigDecimal.ZERO;
    for (TransactionProduct product : products) {
      sum = sum.add(product.getPrice().multiply(BigDecimal.valueOf(product.getQuantity())));
    }
    return sum;
  }

  @JsonProperty("paytype")
  public String getPaytypeName() {
    return paytype.getName();
  }
}

File: src/main/kotlin/ee/leola/kassa/models/Transaction.kt
package ee.leola.kassa.models

import com.fasterxml.jackson.annotation.JsonProperty
import jakarta.persistence.*
import java.time.Instant
import java.time.format.DateTimeFormatter

@Entity
class Transaction(
    var invalid: Boolean = false,
    val created: Instant = Instant.now(),
    @ManyToOne
    val user:  User,
    @OneToMany(cascade = [CascadeType.ALL])
    @JoinTable(
        name = "transaction_transaction_product",
        joinColumns = [JoinColumn(name = "transaction_id", referencedColumnName = "id")],
        inverseJoinColumns = [JoinColumn(name = "transaction_product_id", referencedColumnName = "id")]
    )
    val products: Set<TransactionProduct>,
    @ManyToOne
    private val paytype:  Paytype
) : Model() {
    fun invalidate(): Transaction {
        invalid = true
        return this
    }

    @get:Transient
    val formattedTime get() = DateTimeFormatter.ISO_INSTANT.format(created)

    @get:Transient
    val sum get() = products.sumOf { it.price * it.quantity.toBigDecimal() }

    @get:Transient
    @get:JsonProperty("paytype")
    val paytypeName get() = paytype.name

    fun affectsQuantity() = paytype.affectsQuantity
    fun affectsBalance() = paytype.affectsBalance
}